File: 2024/17.py
import timeit, copy

from utils import readInputWithBlank

# CONSIDERAZIONI SU reminder vs module
# DISASSEMBLER

class IntCode:
    def __init__(self):
        self.A = 0
        self.B = 0
        self.C = 0
        self.code = None
        self.rom = None
        self.pointer = 0
        self.out = []

    def reset(self):
        self.code = copy.deepcopy(self.rom['code'])
        self.A = self.rom['A']
        self.B = self.rom['B']
        self.C = self.rom['C']
        self.pointer = 0
        self.out = []

    def setCode(self, code):
        self.code = list(map(int, code.split(",")))
        self.rom = {"code":copy.deepcopy(self.code), "A":self.A, "B":self.B, "C":self.C}

    def setRegister(self, A=None, B=None, C=None):
        if A is not None:
            self.A = A
        if B is not None:
            self.B = B
        if C is not None:
            self.C = C

    def combo(self, value):
        if value == 7:
            raise ValueError("Reserved value !")
        if value == 4:
            return self.A
        elif value == 5:
            return self.B
        elif value == 6:
            return self.C
        else:
            return value
        
    def output(self):
        return ",".join(map(str, self.out))

    def run(self, debug=False):
        if debug:
            print (f"A: {self.A} B: {self.B} C: {self.C}")
        while self.pointer < len(self.code):
            optcode = self.code[self.pointer]
            operand = self.code[self.pointer+1]
            if optcode == 0:    # adv
                if debug:
                    print ("adv ", operand)
                self.A = self.A//2**self.combo(operand)
                self.pointer += 2
            elif optcode == 1:  # bxl
                if debug:
                    print ("bxl ", operand)
                self.B ^= operand
                self.pointer += 2
            elif optcode == 2:  # bst
                if debug:
                    print ("bst ", operand)
                self.B = self.combo(operand) % 8
                self.pointer += 2
            elif optcode == 3:  # jnz
                if debug:
                    print ("jnz ", operand)
                
                if self.A != 0:
                    self.pointer = operand
                else:
                    self.pointer += 2
            elif optcode == 4:  # bxc
                if debug:
                    print ("bxc ", operand)
                
                self.B = self.B ^ self.C
                self.pointer  += 2
            elif optcode == 5:  # out
                self.out.append(self.combo(operand) % 8)
                if debug:
                    print ("out ", self.out)
                self.pointer  += 2
            elif optcode == 6:  # bdv
                if debug:
                    print ("bdv ", operand)
                
                self.B = self.A//2**self.combo(operand)
                self.pointer  += 2
            elif optcode == 7:  # cdv
                if debug:
                    print ("cdv ", operand)
                
                self.C = self.A//2**self.combo(operand)
                self.pointer  += 2
            if debug:
                input()

def loadInput():
    #lines = readInputWithBlank("input_17_prova.txt")
    lines = readInputWithBlank("input_17.txt")

    intcode = IntCode()
    for l in lines:
        if l == "":
            continue
        if l.startswith("Register A:"):
            intcode.setRegister(A=int(l.split(": ")[1]))
        elif l.startswith("Register B:"):
            intcode.setRegister(B=int(l.split(": ")[1]))
        elif l.startswith("Register C:"):
            intcode.setRegister(C=int(l.split(": ")[1]))
        else:
            intcode.setCode(l.split(": ")[1])
    return intcode

def part1(intcode):
    intcode.run()
    print (f"🎄 Part 1: {intcode.output()}")

def pyintcode(A):
    B = A%8
    B = B^6
    C = A//2**B
    B = B ^ C
    B = B ^ 4
    return B % 8

def findQuin(A, program, col=0, res=[]):
    if pyintcode(A) != program[-(col + 1)]:
        return

    if col == len(program) - 1:
        res.append(A)
    else:
        for B in range(8):
            findQuin(A * 8 + B, program, col + 1, res)

def part2(intcode):
    intcode.reset()
    program = intcode.code
    res = []
    for a in range(8):
        findQuin(a, program, res=res)

    print (f"🎄🎅 Part 2: {res[0]}")

if __name__ == '__main__':
    title = "Day 17: Chronospatial Computer"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()
    
    t1 = timeit.timeit(lambda: part1(inputs), number=1)
    print (f"{t1*1000:.3f} ms")
    
    t2 = timeit.timeit(lambda: part2(inputs), number=1)
    print (f"{t2*1000:.3f} ms")

File: 2024/17.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <chrono>

using namespace std;
using namespace std::chrono;

class IntCode {
private:
    int A;
    int B;
    int C;
    vector<int> code;
    vector<int> out;
    int pointer;

public:
    IntCode() : A(0), B(0), C(0), pointer(0) {}

    void reset() {
        pointer = 0;
        out.clear();
    }

    vector<int> getCode() {
      return code;
    }
    void setCode(const string& code_str) {
        code.clear(); // Clear existing code
        stringstream ss(code_str);
        string token;
        while (getline(ss, token, ',')) {
            code.push_back(stoi(token));
        }
    }

    void setRegister(int a = -1, int b = -1, int c = -1) {
        if (a != -1) A = a;
        if (b != -1) B = b;
        if (c != -1) C = c;
    }

    int getRegister(char reg) const { // Make const
        if (reg == 'A') return A;
        if (reg == 'B') return B;
        if (reg == 'C') return C;
        return -1; // Or throw an exception for invalid register
    }

    string output() const{ // Make const
        string result;
        for (int o : out) {
            result += to_string(o) + ",";
        }
        if (!out.empty()) {
            result.pop_back(); // Remove the trailing comma
        }
        return result;
    }

    void run(bool debug = false) {
        while (pointer < code.size()) {
            int opcode = code[pointer];
            int operand = (pointer + 1 < code.size()) ? code[pointer + 1] : 0;

            if (opcode == 0) { // adv
                A /= (1 << getRegister('B'));
                pointer += 2;
            } else if (opcode == 1) { // bxl
                B ^= operand;
                pointer += 2;
            } else if (opcode == 2) { // bst
                B = getRegister('B') % 8;
                pointer += 2;
            } else if (opcode == 3) { // jnz
                if (A != 0) {
                    pointer = operand;
                } else {
                    pointer += 2;
                }
            } else if (opcode == 4) { // bxc
                B ^= C;
                pointer += 2;
            } else if (opcode == 5) { // out
                out.push_back(getRegister('B') % 8);
                pointer += 2;
            } else if (opcode == 6) { // bdv
                B = A / (1 << getRegister('B'));
                pointer += 2;
            } else if (opcode == 7) { // cdv
                C = A / (1 << getRegister('B'));
                pointer += 2;
            } else {
                cerr << "Invalid opcode: " << opcode << endl;
                return; // Exit the run loop on invalid opcode
            }

            if (debug) {
                cout << "A: " << A << " B: " << B << " C: " << C << endl;
            }
        }
    }
};

IntCode loadInput(const string& filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error opening file: " << filename << endl;
        exit(1);
    }

    IntCode intcode;
    string line;
    while (getline(file, line)) {
        if (line.empty()) continue;
        if (line.find("Register A:") != string::npos) {
            intcode.setRegister(stoi(line.substr(13)), -1, -1);
        } else if (line.find("Register B:") != string::npos) {
            intcode.setRegister(-1, stoi(line.substr(13)), -1);
        } else if (line.find("Register C:") != string::npos) {
            intcode.setRegister(-1, -1, stoi(line.substr(13)));
        } else if (line.find("Code:") != string::npos) {
            intcode.setCode(line.substr(7));
        }
    }
    return intcode;
}

int pyintcode(int A) {
    int B = A % 8;
    B = B ^ 6;
    int C = A / (1 << B);
    B = B ^ C;
    B = B ^ 4;
    return B % 8;
}

void findQuin(int A, const vector<int>& program, int col, vector<int>& res) {
    if (pyintcode(A) != program[program.size() - 1 - col]) {
        return;
    }

    if (col == program.size() - 1) {
        res.push_back(A);
    } else {
        for (int B = 0; B < 8; ++B) {
            findQuin(A * 8 + B, program, col + 1, res);
        }
    }
}

void part1(IntCode& intcode) {
    intcode.run();
    cout << "🎄 Part 1: " << intcode.output() << endl;
}

void part2(IntCode& intcode) {
    intcode.reset();
    const vector<int>& program = intcode.getCode();
    vector<int> res;
    for (int a = 0; a < 8; ++a) {
        findQuin(a, program, 0, res);
    }

    cout << "🎄🎅 Part 2: " << res[0] << endl;
}

int main() {
    string title = "Day 17: Chronospatial Computer";
    string sub = string(title.length() + 2, '-');

    cout << endl;
    cout << " " << title << " " << endl;
    cout << sub << endl;

    IntCode inputs = loadInput("input_17.txt");

    auto t1_start = high_resolution_clock::now();
    part1(inputs);
    auto t1_end = high_resolution_clock::now();
    auto duration1 = duration_cast<milliseconds>(t1_end - t1_start);
    cout << duration1.count() << " ms" << endl;

    auto t2_start = high_resolution_clock::now();
    part2(inputs);
    auto t2_end = high_resolution_clock::now();
    auto duration2 = duration_cast<milliseconds>(t2_end - t2_start);
    cout << duration2.count() << " ms" << endl;

    return 0;
}